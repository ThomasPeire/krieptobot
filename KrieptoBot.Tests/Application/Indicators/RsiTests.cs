using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using FluentAssertions;
using KrieptoBot.Application.Indicators;
using KrieptoBot.Domain.Trading.ValueObjects;
using KrieptoBot.Infrastructure.Bitvavo.Dtos;
using KrieptoBot.Infrastructure.Bitvavo.Extensions;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using Snapshooter.NUnit;

namespace KrieptoBot.Tests.Application.Indicators
{
    internal class RsiTests
    {
        private IEnumerable<Candle> _candles;

        [SetUp]
        public void Setup()
        {
            InitCandles();
        }

        private void InitCandles()
        {
            var candlesJson = File.ReadAllText(@"./MockData/Bitvavo/candles_btc-eur.json");
            var deserializedCandles = JsonConvert.DeserializeObject(candlesJson) as JArray;
            _candles = deserializedCandles.Select(x =>
                new CandleDto
                {
                    TimeStamp = x.Value<long>(0),
                    Open = x.Value<decimal>(1),
                    High = x.Value<decimal>(2),
                    Low = x.Value<decimal>(3),
                    Close = x.Value<decimal>(4),
                    Volume = x.Value<decimal>(5)
                }.ConvertToKrieptoBotModel());
        }

        [Test]
        public void Rsi_Should_CalculateRsi()
        {
            var datetimeFrom = new DateTime(2021, 08, 13);

            var candlesToWorkWith = _candles
                .Where(x => x.TimeStamp >= datetimeFrom && x.TimeStamp <= datetimeFrom.AddDays(1))
                .OrderBy(x => x.TimeStamp);

            var rsiValues = new Rsi().Calculate(candlesToWorkWith, 14);

            rsiValues.Should().MatchSnapshot();
        }
    }
}
