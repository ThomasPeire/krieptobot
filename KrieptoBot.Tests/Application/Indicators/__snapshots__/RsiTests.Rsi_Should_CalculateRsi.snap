    {
  "2021-11-01T00:00:00Z": 0.0,
  "2021-11-01T00:05:00Z": 0.0,
  "2021-11-01T00:10:00Z": 0.0,
  "2021-11-01T00:15:00Z": 0.0,
  "2021-11-01T00:20:00Z": 0.0,
  "2021-11-01T00:25:00Z": 0.0,
  "2021-11-01T00:30:00Z": 0.0,
  "2021-11-01T00:35:00Z": 0.0,
  "2021-11-01T00:40:00Z": 0.0,
  "2021-11-01T00:45:00Z": 0.0,
  "2021-11-01T00:50:00Z": 0.0,
  "2021-11-01T00:55:00Z": 0.0,
  "2021-11-01T01:00:00Z": 0.0,
  "2021-11-01T01:05:00Z": 0.0,
  "2021-11-01T01:10:00Z": 0.0,
  "2021-11-01T01:15:00Z": 67.299225747684833763473508425,
  "2021-11-01T01:20:00Z": 67.522239877521196024078219417,
  "2021-11-01T01:25:00Z": 68.411571299276381613029469724,
  "2021-11-01T01:30:00Z": 57.624690571426659770453715747,
  "2021-11-01T01:35:00Z": 58.063656896491365720402484581,
  "2021-11-01T01:40:00Z": 50.599905527992542972848997483,
  "2021-11-01T01:45:00Z": 52.050182035184340544727315554,
  "2021-11-01T01:50:00Z": 49.751271102596438791029421806,
  "2021-11-01T01:55:00Z": 47.266405814052701288732304686,
  "2021-11-01T02:00:00Z": 46.905981945542796339493895495,
  "2021-11-01T02:05:00Z": 52.786336229776023853968467404,
  "2021-11-01T02:10:00Z": 42.306198360137906054084134356,
  "2021-11-01T02:15:00Z": 25.703080749489717039626474678,
  "2021-11-01T02:20:00Z": 21.329855052360131689694606733,
  "2021-11-01T02:25:00Z": 20.780025976279636307458826257,
  "2021-11-01T02:30:00Z": 26.776513078464332769115807558,
  "2021-11-01T02:35:00Z": 31.640589794993025017284571211,
  "2021-11-01T02:40:00Z": 30.173614352197101673895688422,
  "2021-11-01T02:45:00Z": 30.619011986858971759343653324,
  "2021-11-01T02:50:00Z": 36.306872696636304083231868398,
  "2021-11-01T02:55:00Z": 40.609146937900349199434739592,
  "2021-11-01T03:00:00Z": 36.764738392193108665787401507,
  "2021-11-01T03:05:00Z": 36.388379642410367901175480324,
  "2021-11-01T03:10:00Z": 34.564754926861547196043996072,
  "2021-11-01T03:15:00Z": 33.511952277740189023884304138,
  "2021-11-01T03:20:00Z": 33.232443252744257396441442987,
  "2021-11-01T03:25:00Z": 37.02246128276567531802427866,
  "2021-11-01T03:30:00Z": 34.416151301856523756962853855,
  "2021-11-01T03:35:00Z": 34.595375679391747419680519505,
  "2021-11-01T03:40:00Z": 35.915562764520145151698512731,
  "2021-11-01T03:45:00Z": 33.881417436705301960548537644,
  "2021-11-01T03:50:00Z": 32.745058069413282060307661923,
  "2021-11-01T03:55:00Z": 26.159626467867973474270506398,
  "2021-11-01T04:00:00Z": 24.871495799253600901957039013,
  "2021-11-01T04:05:00Z": 25.588939570500227554015848407,
  "2021-11-01T04:10:00Z": 43.444735817644881050703270679,
  "2021-11-01T04:15:00Z": 41.723787452242165249981160647,
  "2021-11-01T04:20:00Z": 49.043629559266526656900434852,
  "2021-11-01T04:25:00Z": 46.780904632572198339829290385,
  "2021-11-01T04:30:00Z": 57.741750063360719806441523232,
  "2021-11-01T04:35:00Z": 55.890329702690115666151457517,
  "2021-11-01T04:40:00Z": 57.393524996559191693334742431,
  "2021-11-01T04:45:00Z": 54.621206839084071328400185042,
  "2021-11-01T04:50:00Z": 56.13086636023872110689391825,
  "2021-11-01T04:55:00Z": 55.529403716772269258086806739,
  "2021-11-01T05:00:00Z": 52.368390007601411796986189416,
  "2021-11-01T05:05:00Z": 54.928294047973745826961237106,
  "2021-11-01T05:10:00Z": 57.797380819934167796519465659,
  "2021-11-01T05:15:00Z": 56.540303396343968140072569589,
  "2021-11-01T05:20:00Z": 61.164307346311846501567708929,
  "2021-11-01T05:25:00Z": 55.673983587588471071332640326,
  "2021-11-01T05:30:00Z": 54.177622268691695002837756643,
  "2021-11-01T05:35:00Z": 53.151977365544924801233288814,
  "2021-11-01T05:40:00Z": 56.968255170754600577570249618,
  "2021-11-01T05:45:00Z": 52.962847395807047052018140663,
  "2021-11-01T05:50:00Z": 49.467012141465313537552539392,
  "2021-11-01T05:55:00Z": 47.957338601736625654837777544,
  "2021-11-01T06:00:00Z": 44.512306211886402191670212974,
  "2021-11-01T06:05:00Z": 42.060761173969222150918342181,
  "2021-11-01T06:10:00Z": 42.86179657056917348669330465,
  "2021-11-01T06:15:00Z": 39.388216645448653184924502378,
  "2021-11-01T06:20:00Z": 45.44775560942815402497556975,
  "2021-11-01T06:25:00Z": 54.075239187305424614693666349,
  "2021-11-01T06:30:00Z": 56.29526857304658619028619308,
  "2021-11-01T06:35:00Z": 64.280324270066242040616007936,
  "2021-11-01T06:40:00Z": 60.888620702634297728044077057,
  "2021-11-01T06:45:00Z": 53.229408289093848081010811306,
  "2021-11-01T06:50:00Z": 54.100696396052115299016025866,
  "2021-11-01T06:55:00Z": 55.59881837685851736331090836,
  "2021-11-01T07:00:00Z": 70.090153486994998103418486414,
  "2021-11-01T07:05:00Z": 73.077401181825724066660419969,
  "2021-11-01T07:10:00Z": 79.73742284645688143845475733,
  "2021-11-01T07:15:00Z": 77.228389286803123291904771607,
  "2021-11-01T07:20:00Z": 78.605845157673808972473342341,
  "2021-11-01T07:25:00Z": 74.558373589696880011020086296,
  "2021-11-01T07:30:00Z": 75.377584184119587082914937942,
  "2021-11-01T07:35:00Z": 68.274971172357796463697239471,
  "2021-11-01T07:40:00Z": 69.400784768272006849827666245,
  "2021-11-01T07:45:00Z": 70.924165877696292477634571167,
  "2021-11-01T07:50:00Z": 71.950336743733427421027050691,
  "2021-11-01T07:55:00Z": 69.800523540844641822466327855,
  "2021-11-01T08:00:00Z": 71.969078989946066741118016341
}
